openedx_lti_tool_plugin/tests/views/test_launch_view.py

```diff
@@ -0,0 +1,802 @@
+        """Test get_lti_profile method without LTI profile created.
+
+        Args:
+            get_or_create_mock: Mocked LtiProfile.objects get_or_create method.
+        """
+        get_or_create_mock.return_value = self.lti_profile_mock, True
+
+        self.assertEqual(self.view_class().get_lti_profile(ISS, AUD, SUB, PII), self.lti_profile_mock)
+        get_or_create_mock.assert_called_once_with(
+            platform_id=ISS,
+            client_id=AUD,
+            subject_id=SUB,
+            defaults={'pii': PII},
+        )
+        self.lti_profile_mock.update_pii.assert_not_called()
+
+
+@patch('openedx_lti_tool_plugin.views.authenticate')
+@patch('openedx_lti_tool_plugin.views.login')
+@patch('openedx_lti_tool_plugin.views.mark_user_change_as_expected')
+class TestLtiToolLaunchViewAuthenticateAndLogin(TestLtiToolLaunchViewBase):
+    """Testcase for LtiToolLaunchView authenticate_and_login method."""
+
+    def test_authenticate_and_login_with_user(
+        self,
+        mark_user_change_as_expected_mock: MagicMock,
+        login_mock: MagicMock,
+        authenticate_mock: MagicMock,
+    ):
+        """Test authenticate_and_login method with user.
+
+        Args:
+            mark_user_change_as_expected_mock: Mocked mark_user_change_as_expected function.
+            login_mock: Mocked login function.
+            authenticate_mock: Mocked authenticate function.
+        """
+        authenticate_mock.return_value = self.user
+
+        self.assertEqual(self.view_class().authenticate_and_login(None, **BASE_LAUNCH_DATA), self.user)
+        authenticate_mock.assert_called_once_with(None, **BASE_LAUNCH_DATA)
+        login_mock.assert_called_once_with(None, self.user)
+        mark_user_change_as_expected_mock.assert_called_once_with(self.user.id)
+
+    def test_authenticate_and_login_without_user(
+        self,
+        mark_user_change_as_expected_mock: MagicMock,
+        login_mock: MagicMock,
+        authenticate_mock: MagicMock,
+    ):
+        """Test authenticate_and_login method without user.
+
+        Args:
+            mark_user_change_as_expected_mock: Mocked mark_user_change_as_expected function.
+            login_mock: Mocked login function.
+            authenticate_mock: Mocked authenticate function.
+        """
+        authenticate_mock.return_value = None
+
+        with self.assertRaises(LtiToolLaunchException) as ex:
+            self.view_class().authenticate_and_login(None, **BASE_LAUNCH_DATA)
+        self.assertEqual(
+            str(ex.exception),
+            'Profile authentication failed.',
+        )
+        authenticate_mock.assert_called_once_with(None, **BASE_LAUNCH_DATA)
+        login_mock.assert_not_called()
+        mark_user_change_as_expected_mock.assert_not_called()
+
+
+@patch('openedx_lti_tool_plugin.views.course_enrollment')
+class TestLtiToolLaunchViewEnroll(TestLtiToolLaunchViewBase):
```

Comment:
@alexjmpb Does this cases cover when course_enrollment_exception is raised?

Reply:
No I added one